schema/property_filter.go

```diff
@@ -0,0 +1,133 @@
+package schema
+
+import "github.com/pkg/errors"
+
+type FilterFactory struct {
+	visible, hidden []string
+}
+
+func (f *FilterFactory) CreateFilterFromProperties(visible, hidden []string) (*Filter, error) {
+	f.visible = visible
+	f.hidden = hidden
+
+	strategy, err := f.createFilterStrategy()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Filter{filterStrategy: strategy}, nil
+}
+
+func CreateExcludeAllFilter() *Filter {
+	return &Filter{filterStrategy: &excludeAllFilter{}}
+}
+
+func (f *FilterFactory) createFilterStrategy() (FilterStrategy, error) {
+	switch f.getFilterType() {
+	case All:
+		return &includeAllFilter{}, nil
+	case Visible:
+		return &visibleFilter{properties: sliceToSet(f.visible)}, nil
+	case Hidden:
+		return &hiddenFilter{properties: sliceToSet(f.hidden)}, nil
+	default:
+		return nil, errors.New("Cannot have filter with both visible and hidden properties")
+	}
+}
+
+type filterType int
+
+const (
+	All filterType = iota
+	Visible
+	Hidden
+	Invalid
+)
+
+func (f *FilterFactory) getFilterType() filterType {
+	if f.visible == nil {
+		if f.hidden == nil {
+			return All
+		}
+		return Hidden
+	}
+	if f.hidden == nil {
+		return Visible
+	}
+	return Invalid
+}
+
+func sliceToSet(keys []string) map[string]bool {
+	result := make(map[string]bool)
+	for _, key := range keys {
+		result[key] = true
+	}
+	return result
+}
+
+type Filter struct {
+	filterStrategy FilterStrategy
+}
+
+func (f *Filter) RemoveHiddenKeysFromMap(data map[string]interface{}) map[string]interface{} {
+	result := make(map[string]interface{})
+	for key, value := range data {
+		if f.filterStrategy.Filter(key) {
+			result[key] = value
+		}
+	}
+	return result
+}
+
+func (f *Filter) RemoveHiddenKeysFromSlice(data []string) []string {
+	result := make([]string, 0)
+	for _, key := range data {
+		if f.filterStrategy.Filter(key) {
+			result = append(result, key)
+		}
+	}
+	return result
+}
+
+func (f *Filter) IsForbidden(key string) bool {
+	return !f.filterStrategy.Filter(key)
+}
+
+func (f *Filter) AllowsAll() bool {
+	_, ok := f.filterStrategy.(*includeAllFilter)
```

Comment:
closing to initial implementation. Also AllowsAll should take input and validates whether all passes otherwise I don't understand what is should do.

Reply:
AllowsAll checks if given filter allows all keys it is only possible for includeAllFilter.

Reply:
no, this function checks whether given filter is of type includeAllFIlter. Name doesn't explain semantic.

Reply:
There is only one usage of this function: in `Check` in `policy.go` to write a debug message.
Checking all properties here would be redundant, since they are checked again anyway, so I'd rather removed this function completely and also the debug message. What do you think about this solution?

Reply:
Branch we are talking about is not only displaying log message, but also does an early return. I just don't want to rely on includeAllPredicate implementation. AllowsAll should not rely on type casting and believing on that Validate method always returns true. You can remove this early check - will cause small performance drop or change interface to delegate all/any/none predicate checking to predicate. In this solution you will be able preserve function complexity by implementing this semantic to these interface methods. For example includeAllPredicate::All(properties) would always return true no matter what is passed to function.

Reply:
Please check the comment about removing this logic.